Shell/Shell.cpp

```diff
@@ -950,417 +249,232 @@ Vector<String> Shell::expand_globs(const StringView& path, const StringView& bas
             }
-            children.append({ argv[0], child });
-
-            StringBuilder cmd;
-            cmd.join(" ", argv_string);
-
-            auto job = make<Job>(child, (unsigned)child, cmd.build(), find_last_job_id() + 1);
-            jobs.set((u64)child, move(job));
+            _exit(126);
         }
+        ASSERT_NOT_REACHED();
+    }
 
-#ifdef SH_DEBUG
-        dbgprintf("Closing fds in shell process:\n");
-#endif
-        fds.collect();
+    StringBuilder cmd;
+    cmd.join(" ", command.argv);
 
-#ifdef SH_DEBUG
-        dbgprintf("Now we gotta wait on children:\n");
-        for (auto& child : children)
-            dbgprintf("  %d (%s)\n", child.pid, child.name.characters());
-#endif
+    auto job = adopt(*new Job(child, (unsigned)child, cmd.build(), find_last_job_id() + 1));
+    job->on_exit = [&] {
+        dbg() << "Hey, I '" << job->cmd() << "' died";
+        if (job->is_running_in_background())
+            fprintf(stderr, "Shell: Job %d(%s) exited\n", job->pid(), job->cmd().characters());
+        jobs.remove(job->pid());
+    };
+    jobs.set((u64)child, job);
 
-        if (command.attributes & Attributes::InBackground) {
-            // Set the jobs as running in background and continue without waiting.
-            for (auto& child : children)
-                const_cast<Job*>(jobs.get(child.pid).value())->set_running_in_background(true);
+    // FIXME: Figure out why this happens.
+    //        Not doing this causes a race between LibLine's DSR and the child's rewirings.
+    usleep(300);
 
-            continue;
-        }
+    return *job;
+}
 
-        for (size_t i = 0; i < children.size(); ++i) {
-            auto& child = children[i];
-            dbg() << "Now waiting for " << child.name << " (" << child.pid << ")";
-            do {
-                if (wait_for_pid(child, i != children.size() - 1, return_value) == IterationDecision::Break)
-                    break;
-            } while (errno == EINTR);
-        }
+void Shell::take_back_stdin()
+{
+    tcsetpgrp(0, getpid());
+    tcsetattr(0, TCSANOW, &termios);
+}
 
-        if (command.attributes & Attributes::ShortCircuitOnFailure) {
-            if (return_value != 0) {
-                fail_short_circuits = true;
-            }
-        }
-    }
+void Shell::block_on_job(const RefPtr<Job> job_ptr)
+{
+    auto job = const_cast<Job*>(job_ptr.ptr());
+    if (!job)
+        return;
 
-    last_return_code = return_value;
+    Core::EventLoop loop;
+    dbg() << "Trying to wait for '" << job->cmd() << "', it had on_exit? " << !!job->on_exit;
+    job->on_exit = nullptr;
+    dbg() << "JOB EXIT NUKED";
+    job->on_exit = [&] {
+        dbg() << "Hey, I '" << job->cmd() << "' died!";
+        if (job->is_running_in_background())
+            fprintf(stderr, "Shell: Job %d(%s) exited\n", job->pid(), job->cmd().characters());
+        jobs.remove(job->pid());
+        Core::EventLoop::current().quit(0);
+    };
```

Comment:
I am so very confused...somehow the old callback is invoked after the job exits (note the lack of '!' after 'died')
```
Shell(22): Trying to wait for 'cat', it had on_exit? true
Shell(22): JOB EXIT NUKED
Shell(22): Job 'cat' primed, it has quited? false
[Shell(23:23)]: do_exec(/bin/ls)
[Shell(24:24)]: do_exec(/bin/cat)
[FinalizerTask(3:3)]: Unblock Shell(22:22) due to signal
Shell(22): Hey, I 'cat' died
```

Reply:
I poked a little into Function, __how does this even call the thing that's null?__
```
Shell(22): Job on_exit A addr=0x02a500e6                             <- original m_callable_wrapper
Shell(22): Trying to wait for 'cat', it had on_exit? true
Shell(22): JOB EXIT NUKED
Shell(22): Job on_exit B addr=0x00000000                             <- m_callable_wrapper = nullptr
Shell(22): Job 'cat' primed, it has quited? false
[Shell(23:23)]: do_exec(/bin/ls)
[Shell(24:24)]: do_exec(/bin/cat)
[FinalizerTask(3:3)]: Unblock Shell(22:22) due to signal
------------- callback start -------------------
Shell(22): I am 22, the shell was 22
Shell(22): Job on_exit C addr=0x00000000                             <- ...=nullptr inside the callback
Shell(22): Hey, I 'cat' died
```

Reply:
I dunno what was going on there with null callbacks, but moving it directly to a capture worked...
issue sidestepped for now :shrug: